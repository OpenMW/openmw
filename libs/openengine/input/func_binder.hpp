#ifndef OENGINE_INPUT_FUNCBINDER_H
#define OENGINE_INPUT_FUNCBINDER_H

#include <string>
#include <vector>
#include <boost/function.hpp>
#include <cassert>

namespace OEngine {
namespace Input {

/**
   An Action defines the user defined action corresponding to a
   binding.

   The first parameter is the action index that invoked this call. You
   can assign the same function to multiple actions, and this can help
   you keep track of which action was invoked.

   The second parameter is an optional user-defined parameter,
   represented by a void pointer. In many cases it is practical to
   point this to temporaries (stack values), so make sure not to store
   permanent references to it unless you've planning for this on the
   calling side as well.
 */
typedef boost::function<void(int,const void*)> Action;

/**
   The FuncBinder is a simple struct that binds user-defined indices
   to functions. It is useful for binding eg. keyboard events to
   specific actions in your program, but can potentially have many
   other uses as well.
 */
class FuncBinder
{
  struct FuncBinding
  {
    std::string name;
    Action action;
  };

  std::vector<FuncBinding> bindings;

public:
  /**
     Constructor. Initialize the struct by telling it how many action
     indices you intend to bind.

     The indices you use should be 0 <= i < number.
  */
  FuncBinder(int number) : bindings(number) {}

  unsigned int getSize() { return bindings.size(); }

  /**
     Bind an action to an index.
   */
  void bind(int index, Action action, const std::string &name="")
  {
    assert(index >= 0 && index < (int)bindings.size());

    FuncBinding &fb = bindings[index];
    fb.action = action;
    fb.name = name;
  }

  /**
     Unbind an index, reverting a previous bind().
   */
  void unbind(int index)
  {
    assert(index >= 0 && index < (int)bindings.size());

    bindings[index] = FuncBinding();
  }

  /**
     Call a specific action. Takes an optional parameter that is
     passed to the action.
   */
  void call(int index, const void *p=NULL) const
  {
    assert(index >= 0 && index < (int)bindings.size());

    const FuncBinding &fb = bindings[index];
    if(fb.action) fb.action(index, p);
  }

  /// Check if a given index is bound to anything
  bool isBound(int index) const
  {
    assert(index >= 0 && index < (int)bindings.size());

    return !bindings[index].action.empty();
  }

  /// Return the name associated with an action (empty if not bound)
  const std::string &getName(int index) const
  {
    assert(index >= 0 && index < (int)bindings.size());

    return bindings[index].name;
  }
};
}}
#endif
