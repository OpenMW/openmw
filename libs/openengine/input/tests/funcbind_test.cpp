#include <iostream>
using namespace std;

#include "../func_binder.hpp"

void f1(int i, const void *p)
{
  cout << "  F1 i=" << i << endl;

  if(p)
    cout << "  Got a nice gift: "
         << *((const float*)p) << endl;
}

void f2(int i, const void *p)
{
  cout << "  F2 i=" << i << endl;
}

using namespace OEngine::Input;

int main()
{
  cout << "This will test the function binding system\n";

  FuncBinder bnd(5);

  bnd.bind(0, &f1, "This is action 1");
  bnd.bind(1, &f2);
  bnd.bind(2, &f1, "This is action 3");
  bnd.bind(3, &f2, "This is action 4");

  bnd.unbind(2);

  for(int i=0; i<5; i++)
    {
      cout << "Calling " << i << ": '" << bnd.getName(i) << "'\n";
      bnd.call(i);
      if(!bnd.isBound(i)) cout << "  (not bound)\n";
    }

  cout << "\nCalling with parameter:\n";
  float f = 3.1415;
  bnd.call(0, &f);

  return 0;
}
