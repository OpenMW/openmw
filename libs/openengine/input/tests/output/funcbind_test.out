This will test the function binding system
Calling 0: 'This is action 1'
  F1 i=0
Calling 1: ''
  F2 i=1
Calling 2: ''
  (not bound)
Calling 3: 'This is action 4'
  F2 i=3
Calling 4: ''
  (not bound)

Calling with parameter:
  F1 i=0
  Got a nice gift: 3.1415
